File: wakeonlan/src/main/java/com/github/shingyx/wakeonlan/MainActivity.java
package com.github.shingyx.wakeonlan;

import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;

public class MainActivity extends AppCompatActivity {

    private MagicPacketProcessor magicPacketProcessor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        magicPacketProcessor = new MagicPacketProcessor(this);

        EditText macAddressField = findViewById(R.id.macAddress);
        Button sendButton = findViewById(R.id.send);
        ProgressBar progressBar = findViewById(R.id.progressBar);

        macAddressField.setText(magicPacketProcessor.getSavedMacAddress());

        sendButton.setOnClickListener(v -> {
            progressBar.setVisibility(View.VISIBLE);

            String macAddress = macAddressField.getText().toString().trim();

            new SendPacketTask(magicPacketProcessor, macAddress, (String error) -> {
                if (error == null) {
                    magicPacketProcessor.setSavedMacAddress(macAddress);
                }
                this.runOnUiThread(() -> {
                    progressBar.setVisibility(View.INVISIBLE);

                    new AlertDialog.Builder(this)
                            .setTitle(error == null ? R.string.packet_sent : R.string.error)
                            .setMessage(error)
                            .setPositiveButton(R.string.ok, null)
                            .create()
                            .show();
                });
            }).execute();
        });
    }
}

File: wakeonlan/src/main/java/com/github/shingyx/wakeonlan/MainActivity.kt
package com.github.shingyx.wakeonlan

import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
    private val scope = MainScope()

    private lateinit var magicPacketProcessor: MagicPacketProcessor

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        magicPacketProcessor = MagicPacketProcessor(this)

        macAddressField.setText(magicPacketProcessor.getSavedMacAddress())

        sendButton.setOnClickListener {
            scope.launch {
                progressBar.visibility = View.VISIBLE

                val macAddress = macAddressField.text.toString().trim()
                val result = Result.runCatching {
                    magicPacketProcessor.send(macAddress)
                    magicPacketProcessor.saveMacAddress(macAddress)
                }
                val error = result.exceptionOrNull()?.message

                progressBar.visibility = View.INVISIBLE

                AlertDialog.Builder(this@MainActivity)
                    .setTitle(if (error == null) R.string.packet_sent else R.string.error)
                    .setMessage(error)
                    .setPositiveButton(R.string.ok, null)
                    .create()
                    .show()
            }
        }
    }
}
